src/lib.rs

```diff
@@ -0,0 +1,1036 @@
+    pub const MAX_BITS: usize = 20;
+    pub const MAX: usize = (1 << Self::MAX_BITS) - 1;
+
+    #[inline(always)]
+    pub fn new(virt_reg: usize, class: RegClass) -> Self {
+        assert!(virt_reg <= Self::MAX);
+        VReg(((virt_reg as u32) << 1) | (class as u8 as u32))
+    }
+
+    #[inline(always)]
+    pub fn vreg(self) -> usize {
+        (self.0 >> 1) as usize
+    }
+
+    #[inline(always)]
+    pub fn class(self) -> RegClass {
+        match self.0 & 1 {
+            0 => RegClass::Int,
+            1 => RegClass::Float,
+            _ => unreachable!(),
+        }
+    }
+
+    #[inline(always)]
+    pub fn invalid() -> Self {
+        VReg::new(Self::MAX, RegClass::Int)
+    }
+}
+
+impl std::fmt::Debug for VReg {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(
+            f,
+            "VReg(vreg = {}, class = {:?})",
+            self.vreg(),
+            self.class()
+        )
+    }
+}
+
+impl std::fmt::Display for VReg {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "v{}", self.vreg())
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct SpillSlot(u32);
+
+impl SpillSlot {
+    #[inline(always)]
+    pub fn new(slot: usize, class: RegClass) -> Self {
+        assert!(slot < (1 << 24));
+        SpillSlot((slot as u32) | (class as u8 as u32) << 24)
+    }
+    #[inline(always)]
+    pub fn index(self) -> usize {
+        (self.0 & 0x00ffffff) as usize
+    }
+    #[inline(always)]
+    pub fn class(self) -> RegClass {
+        match (self.0 >> 24) as u8 {
+            0 => RegClass::Int,
+            1 => RegClass::Float,
+            _ => unreachable!(),
+        }
+    }
+    #[inline(always)]
+    pub fn plus(self, offset: usize) -> Self {
+        SpillSlot::new(self.index() + offset, self.class())
+    }
+
+    #[inline(always)]
+    pub fn invalid() -> Self {
+        SpillSlot(0xffff_ffff)
+    }
+    #[inline(always)]
+    pub fn is_invalid(self) -> bool {
+        self == Self::invalid()
+    }
+    #[inline(always)]
+    pub fn is_valid(self) -> bool {
+        self != Self::invalid()
+    }
+}
+
+impl std::fmt::Display for SpillSlot {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "stack{}", self.index())
+    }
+}
+
+/// An `Operand` encodes everything about a mention of a register in
+/// an instruction: virtual register number, and any constraint/policy
+/// that applies to the register at this program point.
+///
+/// An Operand may be a use or def (this corresponds to `LUse` and
+/// `LAllocation` in Ion).
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub struct Operand {
+    /// Bit-pack into 32 bits. Note that `policy` overlaps with `kind`
+    /// in `Allocation` and we use mutually disjoint tag-value ranges
+    /// so that clients, if they wish, can track just one `u32` per
+    /// register slot and edit it in-place after allocation.
```

Comment:
It would be nice to free up additional tag values for future use. I already have two in mind:
- `ColdAny`: similar to `Any` except that it has lower priority when allocating registers for spilled bundles. The intended use is for instructions that may trap (e.g. a load which might be null) where the trap handler requires certain values to be live (in a register or on the stack). Since the vast majority of loads don't trap, this should be assume to have a reload cost of 0.
- `ExplicitStack`: an explicit stack slot, as we've discussed before.

Reply:
I agree that having a few more tags would be useful here, and also the `OperandOrAllocation` abstraction is no longer useful with the latest Cranelift adaptation so the disjoint tag values were not really serving a purpose anymore. So, updated to no longer require this overlapping.